### PR TITLE
allow collectives to be DCEd during collective optimizations, fix bad partitioner save decision

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/146693. There are more details in the issue discussion.

I tested the repro and confirmed that we no longer save the allgather'd tensor for backward. Going to move a (smaller version of) that repro into a test and update soon.

Here's a tlparse pair of the generated graphs in the test case I wrote, with and without the config:

without config https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/hirsheybar/86f0cbaf-ba47-4896-941b-d3ea75ac85f0/custom/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000

with config: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/hirsheybar/0ad9a0ea-6413-4195-9373-572977896e29/custom/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000

You can see that without the config, we save `wait_tensor_7` for backward. With it, we save `primals_1` for backward and compute the allgather directly in the bw graph

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150302

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @kwen2501 @c-p-i-o

### CODE REVIEW

File: test/inductor/test_distributed_patterns.py [hunk 1]
```diff
@@ -118,6 +120,32 @@ def steps(m, inp):
     return out
 
 
+fw_graph = [None]
+bw_graph = [None]
+
+
+def aot_graph_capture_backend(gm, args):
```
Comment (wconstab) [inline test/inductor/test_distributed_patterns.py:127]: nit: maybe make this a class so 'fw_graph' can be an instance member instead of a global?

File: test/inductor/test_distributed_patterns.py [hunk 2]
```diff
@@ -118,6 +120,32 @@ def steps(m, inp):
...
+        bw_graph[0] = gm
+        return gm
+
+    return aot_module_simplified(
+        gm,
+        args,
+        fw_compiler,
+        bw_compiler,
+        partition_fn=min_cut_rematerialization_partition,
+        keep_inference_input_mutations=True,
```
Comment (wconstab) [inline test/inductor/test_distributed_patterns.py:145]: if these are non-default options, can you put comments as to why we want them in particular in the distributed tests?

File: test/inductor/test_distributed_patterns.py [hunk 3]
```diff
@@ -499,6 +527,209 @@ def test_fake_distributed_inductor(self):
...
+        )
+        ffn(in_, w1, w2)
+        # The main thing to look for here:
+        #  primals_1 is saved for backward.
+        #  in the backward graph, we all_gather(primals_1),
+        #  instead of saving its result for backward
+        self.assertExpectedInline(
+            str(fw_graph[0].code.strip()),
+            """\
+def forward(self, primals_1, primals_2, primals_3):
```
Comment (wconstab) [inline test/inductor/test_distributed_patterns.py:683]: hmm, is it safe to match such a large block of text?  I would be sort of surprised if we can rely on all of this exactly matching (e.g. small naming, ordering changes could happen)

1) wonder if for the fw graph, we just want to (a) identify 'first argument after self' by name, (b) verify that argument is returned from the graph, and then 
2) in the bw graph, we can add assertions that the same var as we returned from fw is being passed into all_gather?

File: test/inductor/test_distributed_patterns.py [hunk 4]
```diff
@@ -499,6 +527,209 @@ def test_fake_distributed_inductor(self):
...
+                return grad_a, grad_b, None
+
+        @torch.compile(backend=aot_graph_capture_backend)
+        def ffn(x: torch.Tensor, w1: torch.Tensor, w2: torch.Tensor) -> torch.Tensor:
+            x = Fp8LinearFn.apply(x, w1, "col")
+            x = torch.nn.functional.relu(x)
+            x = Fp8LinearFn.apply(x, w2, "row")
+            return x
+
+        in_ = torch.randn(
```
Comment (wconstab) [inline test/inductor/test_distributed_patterns.py:660]: hmm, is input sharded on 'batch' dim? some comments about the sharding scheme might be nice.    What's a little odd iiuc is WORLD_SIZE is 2 and represents both DP and TP dims?  So the intent is input sharding on DP dim and w1/w2 sharding on TP dim row/colwise, but the test kinda ignores that and just uses one dim since it's fake anyway?

File: test/inductor/test_distributed_patterns.py [hunk 5]
```diff
@@ -499,6 +527,209 @@ def test_fake_distributed_inductor(self):
...
+            second_t: torch.Tensor,
+            amax_second_t: torch.Tensor,
+            parallel: str,
+        ) -> torch.Tensor:
+            tp_group = torch.distributed.group.WORLD
+            first_fp8, scale_first = scale(first, amax_first)
+            second_t_fp8, scale_second_t = scale(second_t, amax_second_t)
+
+            if parallel == "col":
+                first_fp8 = funcol.all_gather_tensor(
```
Comment (wconstab) [inline test/inductor/test_distributed_patterns.py:581]: related to the assumption about sharding below, iiuc input is 'data parallel' and this all-gather is to create a 'replicated' input?  Then I think it makes sense that we scale the whole input but only use our chunk/scale for Weight.

**Review by wconstab — COMMENTED** [review]
LGTM, mostly nits, although I think its worth fixing / improving the test matching part in case you agree about it being overly sensitive

File: test/inductor/test_distributed_patterns.py (hunk 2 above)
Comment (bdhirsh) [reply test/inductor/test_distributed_patterns.py:145]: yeah these are all the same defaults that vanilla torch.compile uses (annoying boilerplate though)

File: test/inductor/test_distributed_patterns.py (hunk 4 above)
Comment (bdhirsh) [reply test/inductor/test_distributed_patterns.py:660]: hmm to be fair, Luca's [repro](https://github.com/pytorch/pytorch/issues/146693) doesn't hardcode world_size as 2 (it uses `torch.cuda.device_count()`, so I imagine it runs with `8` on an H100).

I'm not totally sure what parallelisms Luca's repro is intended to be using (although it does appear to be issuing allgathers on the input on the forward, and on both weights in the backward). The main thing I confirmed was just the fact that we allgather a weight in the backward, and compile shouldn't save that allgathered weight.

I'm not sure if I should beef up the test to be more explicit about what parallelism is being tested (I'm not entirely sure) - but I can mention in the comment that the parallelism here doesn't really matter, and the "allgather a weight in bw" is the important behavior we're testing if that sounds good to you.

File: test/inductor/test_distributed_patterns.py (hunk 3 above)
Comment (bdhirsh) [reply test/inductor/test_distributed_patterns.py:683]: I mainly did it because it's easy to do with expecttests, and we have some precedent in the codebase. I agree it's a bit of overkill here though - I can probably just assert on e.g. the return string containing the `primal_1` output I want - happy to switch to that (and also that the backward contains the line where we allgather `primal_1` like you're pointing out.

Comment (wconstab) [reply test/inductor/test_distributed_patterns.py:683]: Ok. I'm mainly trying to avoid the test being over sensitive and causing false positives if some subtle thing changes in the graph. But if we are typically asserting based on graphs like this then this is probably fine.

File: test/inductor/test_distributed_patterns.py (hunk 4 above)
Comment (wconstab) [reply test/inductor/test_distributed_patterns.py:660]: The lack of specificity on device mesh dimension just threw me off. Does Luca's example just use the 8gpus directly as a world group or does it create subgroups?  It kinda sounds like he is just doing tensor parallelism, but then I'm confused about what looked like input batch sharding.

**Review by wconstab — APPROVED** [review]

File: test/inductor/test_distributed_patterns.py (hunk 4 above)
Comment (bdhirsh) [reply test/inductor/test_distributed_patterns.py:660]: hmm his example just created a single process group with `device_count` world_size (linked in the github issue). I am not really sure :/

File: torch/distributed/_functional_collectives.py [hunk 6]
```diff
@@ -1041,6 +1041,20 @@ def _reduce_scatter_tensor_coalesced_native_meta(
...
+def treat_collectives_as_safe_to_dce():
+    torch.fx.node._side_effectful_functions.remove(
+        torch.ops._c10d_functional.wait_tensor.default
+    )
+    try:
+        yield
+    finally:
+        torch.fx.node._side_effectful_functions.add(
+            torch.ops._c10d_functional.wait_tensor.default
+        )
```
Comment (zou3519) [inline torch/distributed/_functional_collectives.py:1055]: nesting this context manager will lead to some weird things. That being said you aren't nesting it but it shouldn't be too hard to get it to work? (e.g. only add back what you removed if it was removed)

**Review by zou3519 — APPROVED** [review]

**Comment by malfet** [comment]
@bdhirsh , @ezyang this PR base is 4k+ commits behind trunk

**Comment by ezyang** [comment]
I know, I'll take care of rebasing it

**Comment by ezyang** [comment]
commandeering
